### DISCUSSION TITLE
@apostrophecms/file - file upload error : File type was not accepted. Allowed extensions: *

### DISCUSSION THREAD

**Discussion by ezzle** [apostrophecms/apostrophe#4041] [Q&A]
Hi.

I get the following error when trying to upload a .xml, .png, .odt, .pdf  file : 
`File type was not accepted. Allowed extensions: *`

How to enable .xml extension for upload?

Best regards

Tshitshi

**Comment by BoDonkey** [comment]
Hi @ezzle,
That is a little odd unless you have changed the default attachments `fileGroups` option. Both `.png` and `.pdf` files are allowed by default. This section of the [documentation](https://v3.docs.apostrophecms.org/reference/modules/attachment.html) outlines how to change what files are accepted for upload. Note that just adding an image type doesn't mean that it will be processed, just that it will be written to the uploads folder and written to the database. Let me know if you need more assistance.
Cheers,
Bob

**Reply by ezzle** [reply]
Thank you Bob.

Unfortunately Still not working.

I am trying to upload files via the '@apostrophecms/file' module on the home page.

I am running the project in a GitHub codespace. Does that make a difference?

Not able to find what module throws the error message.

Best regards

Tshitshi

**Reply by BoDonkey** [reply]
At first blush, I would say that codespaces shouldn't make a difference. That message is thrown when the `@apostrophecms/attachment` module doesn't find a file extension in the array of allowed types.
Are you able to upload any files?
I haven't played with codespaces that much, but I'm wondering if you need another volume for the uploads? Based on the error it doesn't make sense.
You can check the `fileGroups` by creating a project level `modules/@apostrophecms/attachment/index.js` file. Then `console.log(self.fileGroups)` from the module `init(self)`. That will let you know what extensions are allowed. If your disallowed file extensions are there, the problem is elsewhere.
Cheers,
Bob

**Reply by ezzle** [reply]
Not able to upload any files. 
Console.log result 
`[
  {
    name: 'images',
    label: 'apostrophe:images',
    extensions: [ 'gif', 'jpg', 'png', 'svg', 'webp' ],
    extensionMaps: { jpeg: 'jpg' }
  },
  {
    name: 'office',
    label: 'apostrophe:office',
    extensions: [
      'txt',  'rtf',  'pdf',
      'xls',  'ppt',  'doc',
      'pptx', 'sldx', 'ppsx',
      'potx', 'xlsx', 'xltx',
      'csv',  'docx', 'dotx'
    ],
    extensionMaps: {}
  },
  { name: 'xml', extensions: [ 'xml', 'pdf' ], extensionMaps: {} }
]`

When trying to upload an image from a page with an image widget, I see these error messages in the browser console
`blob:https://ezzle-didactic-space-funicular-j959wpx6ppcpw69-3000.preview.app.github.dev/875631b1-1dbf-4f2c-9cb8-c4b191de9bb7:1          GET blob:https://ezzle-didactic-space-funicular-j959wpx6ppcpw69-3000.preview.app.github.dev/875631b1-1dbf-4f2c-9cb8-c4b191de9bb7 net::ERR_TIMED_OUT 200 (OK)
Image (asynchrone)
createPlaceholder @ AposModal.vue:265
uploadMedia @ AposModal.vue:213
await in uploadMedia (asynchrone)
invokeWithErrorHandling @ vue.runtime.esm.js:4367
invoker @ vue.runtime.esm.js:3166
original_1._wrapper @ vuedraggable.umd.js:17
AposEditorMixin.js:197          POST https://ezzle-didactic-space-funicular-j959wpx6ppcpw69-3000.preview.app.github.dev/api/v1/@apostrophecms/attachment/upload net::ERR_TIMED_OUT`

**Reply by BoDonkey** [reply]
I think I need to play around with codespace a bit to fully understand those errors. Can you provide the schema field/code that you are using for the upload?

**Reply by BoDonkey** [reply]
One more question. Are you running your codespaces from VSCode locally, or the browser-based version? They seem to act differently.

**Reply by ezzle** [reply]
> I think I need to play around with codespace a bit to fully understand those errors. Can you provide the schema field/code that you are using for the upload?

No particular schema. I am just trying to add a new in the file collection from the Home page. I played with jpg png pdf xml. No success.

**Reply by ezzle** [reply]
> One more question. Are you running your codespaces from VSCode locally, or the browser-based version? They seem to act differently.

Browser based version.

**Reply by BoDonkey** [reply]
I'm having a hard time getting the browser version working nicely. In VSCode it starts up, I can upload files, and they persist between codespaces start-ups. I think we need to look at a repo of your project. Something odd is going on, potentially with other code in the project, that is preventing upload.

**Reply by ezzle** [reply]
Update:
Very odd: 
- Via **Images** collection on the home page, I **can** upload and save an image.
- Via **Files** collection on the home page, I **cannot** upload the same image and get the error: `File type was not accepted. Allowed extensions: *`

It seems the problem is not about codespaces

**Reply by ezzle** [reply]
**### The Problem is limited to the `Files` collection.**
Defining project-specific piece-type for testing, I am able to upload files with any allowed extension.
The defined module has 3 attachment fields : `module.exports = {
	"extend": "@apostrophecms/piece-type",
	"instantiate": true,
	"options": {
		"label": "Essential ReportXml",
		"pluralLabel": "Essential ReportXml",
		"alias": "essential-reportxml"
	},
	fields: {
		add: {
			image: {
				type: 'attachment',
				label: 'esse-image',
				fileGroup: 'images',
				required: true
			},
			xml: {
				type: 'attachment',
				label: 'esse-xml-file',
				fileGroup: 'xml',
				required: true
			},
			file: {
				type: 'attachment',
				label: 'esse-allowed-file',
				required: true
			},
		}
	}
}`

**Reply by ezzle** [reply]
### Additional testing : update file module to extend '@apostrophecms/image'
`module.exports = {
  extend: '@apostrophecms/image',
  options: {
    name: '@apostrophecms/file',
    label: 'apostrophe:file', ...`

I am able to upload files. But the save action produces an error

**Comment by boutell** [comment]
XML is not on our approved list of extensions for attachments. There are
configuration options that can change that. See:

https://v3.docs.apostrophecms.org/reference/modules/attachment.html#options

This is the immediate problem you're having. It's also true as Bob says
that your attachments need to land in a persistent volume of some sort or
they won't "stick." Not sure if Codespaces provides a persistent disk or
not.


On Fri, Feb 10, 2023 at 2:14 PM Robert Means ***@***.***>
wrote:

> At first blush, I would say that codespaces shouldn't make a difference.
> That message is thrown when the @apostrophecms/attachment module doesn't
> find a file extension in the array of allowed types.
> Are you able to upload any files?
> I haven't played with codespaces that much, but I'm wondering if you need
> another volume for the uploads? Based on the error it doesn't make sense.
> You can check the fileGroups by creating a project level
> ***@***.***/attachment/index.js file. Then
> console.log(self.fileGroups) from the module init(self). That will let
> you know what extensions are allowed. If your disallowed file extensions
> are there, the problem is elsewhere.
> Cheers,
> Bob
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apostrophecms/apostrophe/discussions/4041#discussioncomment-4938720>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAH27MCB3RQX3RAQX4PAATWW2HTDANCNFSM6AAAAAAUXVSLDY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: <apostrophecms/apostrophe/repo-discussions/4041/comments/4938720@
> github.com>
>


-- 

THOMAS BOUTELL | CHIEF TECHNOLOGY OFFICER
APOSTROPHECMS | apostrophecms.com | he/him/his

**Comment by boutell** [comment]
I think Bob is probably right, but just for clarification, have you tried a
simple .jpg file? Just making sure things Apostrophe definitely accepts are
not working.

On Sat, Feb 11, 2023 at 6:20 AM Robert Means ***@***.***>
wrote:

> I think I need to play around with codespace a bit to fully understand
> those errors. Can you provide the schema field/code that you are using for
> the upload?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apostrophecms/apostrophe/discussions/4041#discussioncomment-4942677>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAH27N6DNWAHQAK6NX2CZLWW5YZHANCNFSM6AAAAAAUXVSLDY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>


-- 

THOMAS BOUTELL | CHIEF TECHNOLOGY OFFICER
APOSTROPHECMS | apostrophecms.com | he/him/his

**Reply by ezzle** [reply]
> I think Bob is probably right, but just for clarification, have you tried a simple .jpg file? Just making sure things Apostrophe definitely accepts are not working.
> […](#)
> On Sat, Feb 11, 2023 at 6:20 AM Robert Means ***@***.***> wrote: I think I need to play around with codespace a bit to fully understand those errors. Can you provide the schema field/code that you are using for the upload? — Reply to this email directly, view it on GitHub <[#4041 (reply in thread)](https://github.com/apostrophecms/apostrophe/discussions/4041#discussioncomment-4942677)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAH27N6DNWAHQAK6NX2CZLWW5YZHANCNFSM6AAAAAAUXVSLDY> . You are receiving this because you commented.Message ID: ***@***.*** com>
> -- THOMAS BOUTELL | CHIEF TECHNOLOGY OFFICER APOSTROPHECMS | apostrophecms.com | he/him/his

Yes I tried jpg png pdf xml ...

**Reply by ezzle** [reply]
Update:
Very odd: 
- Via **Images** collection on the home page, I **can** upload and save an image.
- Via **Files** collection on the home page, I **cannot** upload the same image and get the error: `File type was not accepted. Allowed extensions: *`

It seems the problem is not about codespaces

**Reply by ezzle** [reply]
From the file/index.js
`// A subclass of `@apostrophecms/piece-type`, `@apostrophecms/file` establishes a library
// of uploaded files, which may be of any type acceptable to the
// [@apostrophecms/attachment](../@apostrophecms/attachment/index.html) module.
// Together with [@apostrophecms/file-widget](../@apostrophecms/file-widget/index.html),
// this module provides a simple way to add downloadable PDFs and the like to
// a website, and to manage a library of them for reuse.`

But  I don't see `@apostrophecms/file-widget` among the modules

**Reply by ezzle** [reply]
**### The Problem is limited to the `Files` collection.**
Defining project-specific piece-type for testing, I am able to upload files with any allowed extension.
The defined module has 3 attachment fields : `module.exports = {
	"extend": "@apostrophecms/piece-type",
	"instantiate": true,
	"options": {
		"label": "Essential ReportXml",
		"pluralLabel": "Essential ReportXml",
		"alias": "essential-reportxml"
	},
	fields: {
		add: {
			image: {
				type: 'attachment',
				label: 'esse-image',
				fileGroup: 'images',
				required: true
			},
			xml: {
				type: 'attachment',
				label: 'esse-xml-file',
				fileGroup: 'xml',
				required: true
			},
			file: {
				type: 'attachment',
				label: 'esse-allowed-file',
				required: true
			},
		}
	}
}`

**Reply by ezzle** [reply]
### Additional testing : update file module to extend '@apostrophecms/image'
`module.exports = {
  extend: '@apostrophecms/image',
  options: {
    name: '@apostrophecms/file',
    label: 'apostrophe:file', ...`

I am able to upload files. But the save action produces an error

**Reply by BoDonkey** [reply]
Edited: This was a comment directed at your earlier code comment.
As it turns out, that comment is a hold-over from a2 if I'm not mistaken. There is no file widget. Instead, files are uploaded through the `attachment` field type. Do you mean the `Files` and `Images` menu items in the admin-bar? If you could put up a repo that would cut down on the guessing a bit.

**Reply by ezzle** [reply]
> Edited: This was a comment directed at your earlier code comment. As it turns out, that comment is a hold-over from a2 if I'm not mistaken. There is no file widget. Instead, files are uploaded through the `attachment` field type. Do you mean the `Files` and `Images` menu items in the admin-bar? If you could put up a repo that would cut down on the guessing a bit.

Yes. Files and images menu items in the admin-bar.

**Reply by ezzle** [reply]
HI.

**Reply by ezzle** [reply]
Hi.
I am still having the same issue with 
`    "apostrophe": "^4.24.1",
`

**Any update ?**

With the recommanded  option to add a new filegroup defined in ..../modules/@apostrophecms/attachment/index.js

`export default {
  options: {
    addFileGroups: [
      // adds xml extension to the 'xsential' group
      {
        name: 'xsential',
        extensions: [
          'xml'
        ],
        image: false
      }
    ]
  },
};`

I tried 2 main cases : 

**1° File type extension**

`export default {
  extend: '@apostrophecms/file',
  options: {
    label: 'Knowledge Base',
    alias: 'xsential-file'
    // Additionally add a `pluralLabel` option if needed.
  },
  fields: {
    add: {
      attachment: {
        label: 'File',
        type: 'attachment',
        fileGroup: 'xsential'
      }
    },
    group: {}
  }
};`

**1.1 Trying to to create a new Knowledge base file**

I can select among other types but can not upload xml files

_UI Error Message `File type was not accepted. Allowed extensions: *`
No Error message on the server console_

**1.2 Trying to create a new Media file**

I can select among other types but can not upload xml files

_UI Error Message `File type was not accepted. Allowed extensions: *`
With the following Error message on the server console :_

`@apostrophecms/file: api-error: BSON field 'distinctCommandRequest.key' is missing but a required field
 {
  module: '@apostrophecms/file',
  type: 'api-error',
  severity: 'error',
  url: '/api/v1/@apostrophecms/file?aposMode=draft&archived=false&viewContext=manage&choices%5B0%5D=_tags&page=1&withPublished=1&aposLocale=en',
  path: '/api/v1/@apostrophecms/file',
  method: 'GET',
  ip: '::1',
  query: {
    aposMode: 'draft',
    archived: 'false',
    viewContext: 'manage',
    choices: [ '_tags' ],
    page: '1',
    withPublished: '1',
    aposLocale: 'en'
  },
  requestId: 't07csi18py4rrespzf75i2c8',
  name: 'error',
  status: 500,
  stack: [
    'at Connection.sendCommand (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/mongodb/lib/cmap/connection.js:306:27)',
    'at process.processTicksAndRejections (node:internal/process/task_queues:103:5)',
    'at async Connection.command (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/mongodb/lib/cmap/connection.js:334:26)',
    'at async Server.command (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/mongodb/lib/sdam/server.js:194:29)',
    'at async tryOperation (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/mongodb/lib/operations/execute_operation.js:213:32)',
    'at async executeOperation (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/mongodb/lib/operations/execute_operation.js:78:16)',
    'at async Collection.distinct (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/mongodb/lib/collection.js:526:16)',
    'at async /Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/doc-type/index.js:2944:26',
    'at async /Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/schema/index.js:1462:23',
    'at async Object.toChoices (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/doc-type/index.js:3000:25)',
    'at async Object.after (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/doc-type/index.js:2794:33)',
    'at async Object.after (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/doc-type/index.js:3395:15)',
    'at async /Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/doc-type/index.js:3122:15',
    'at async getAll (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/piece-type/index.js:277:29)',
    'at async Array.<anonymous> (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/module/index.js:193:30)'
  ],
  cause: undefined,
  errorPath: undefined,
  data: {}
}
` 

**2° Piece type extension with an attachment field**

`export default {
  extend: '@apostrophecms/piece-type',
  options: {
    label: 'Knowledge Base',
    alias: 'xsential-file'
    // Additionally add a `pluralLabel` option if needed.
  },
  fields: {
    add: {
      attachment: {
        label: 'File',
        type: 'attachment',
        fileGroup: 'xsential'
      }
    },
    group: {}
  }
};`

I can select **only** xml files , but  got an error on upload: 

_UI Error `An error occured`
Server console error message :_

`@apostrophecms/attachment: api-error: Cannot read properties of undefined (reading 'xml')
 {
  module: '@apostrophecms/attachment',
  type: 'api-error',
  severity: 'error',
  url: '/api/v1/@apostrophecms/attachment/upload?aposLocale=en',
  path: '/api/v1/@apostrophecms/attachment/upload',
  method: 'POST',
  ip: '::1',
  query: { aposLocale: 'en' },
  requestId: 'k17tfm6ykjnuhk4nw9hkui6u',
  name: 'error',
  status: 500,
  stack: [
    'at /Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/attachment/index.js:555:48',
    'at baseFindIndex (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/lodash/lodash.js:817:11)',
    'at findIndex (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/lodash/lodash.js:7353:14)',
    'at lodash.find (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/lodash/lodash.js:5152:21)',
    'at Object.getFileGroup (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/attachment/index.js:554:18)',
    'at Object.insert (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/attachment/index.js:411:28)',
    'at post.upload (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/attachment/index.js:175:45)',
    'at Array.<anonymous> (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/module/index.js:193:36)',
    'at next (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/apostrophe/modules/@apostrophecms/module/index.js:134:31)',
    'at done (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/multer/lib/make-middleware.js:69:7)',
    'at indicateDone (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/multer/lib/make-middleware.js:73:68)',
    'at /Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/multer/lib/make-middleware.js:232:11',
    'at WriteStream.<anonymous> (/Users/macbookairtknt/0 workspace/xsential/apos-esse/node_modules/multer/storage/disk.js:42:9)',
    'at WriteStream.emit (node:events:520:35)',
    'at finish (node:internal/streams/writable:953:10)',
    'at node:internal/streams/writable:934:13'
  ],
  cause: undefined,
  errorPath: undefined,
  data: {}
}
`
